File: free-ethereum-core/src/main/java/org/ethereum/db/PeerSource.java
/*
 * The MIT License (MIT)
 *
 * Copyright 2017 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 * Copyright (c) [2016] [ <ether.camp> ]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package org.ethereum.db;

import org.apache.commons.lang3.tuple.Pair;
import org.ethereum.datasource.DataSourceArray;
import org.ethereum.datasource.ObjectDataSource;
import org.ethereum.datasource.Serializer;
import org.ethereum.datasource.Source;
import org.ethereum.datasource.leveldb.LevelDbDataSource;
import org.ethereum.net.rlpx.Node;
import org.ethereum.util.RLP;
import org.ethereum.util.RLPList;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.math.BigInteger;

/**
 * Source for {@link org.ethereum.net.rlpx.Node} also known as Peers
 */
public class PeerSource {
    private static final Logger logger = LoggerFactory.getLogger("db");
    private static final Serializer<Pair<Node, Integer>, byte[]> NODE_SERIALIZER = new Serializer<Pair<Node, Integer>, byte[]>() {

        @Override
        public byte[] serialize(final Pair<Node, Integer> value) {
            final byte[] nodeRlp = value.getLeft().getRLP();
            final byte[] nodeIsDiscovery = RLP.encodeByte(value.getLeft().isDiscoveryNode() ? (byte) 1 : 0);
            final byte[] savedReputation = RLP.encodeBigInteger(BigInteger.valueOf(value.getRight()));

            return RLP.encodeList(nodeRlp, nodeIsDiscovery, savedReputation);
        }

        @Override
        public Pair<Node, Integer> deserialize(final byte[] bytes) {
            if (bytes == null) return null;

            final RLPList nodeElement = (RLPList) RLP.decode2(bytes).get(0);
            final byte[] nodeRlp = nodeElement.get(0).getRLPData();
            final byte[] nodeIsDiscovery = nodeElement.get(1).getRLPData();
            final byte[] savedReputation = nodeElement.get(2).getRLPData();
            final Node node = new Node(nodeRlp);
            node.setDiscoveryNode(nodeIsDiscovery != null);

            return Pair.of(node, savedReputation == null ? 0 : (new BigInteger(1, savedReputation)).intValue());
        }
    };
    private final Source<byte[], byte[]> src;
    private DataSourceArray<Pair<Node, Integer>> nodes;


    public PeerSource(final Source<byte[], byte[]> src) {
        this.src = src;
        PeerSource INST = this;
        this.nodes = new DataSourceArray<>(
                new ObjectDataSource<>(src, NODE_SERIALIZER, 512));
    }

    public DataSourceArray<Pair<Node, Integer>> getNodes() {
        return nodes;
    }

    public void clear() {
        if (src instanceof LevelDbDataSource) {
            ((LevelDbDataSource) src).reset();
            this.nodes = new DataSourceArray<>(
                    new ObjectDataSource<>(src, NODE_SERIALIZER, 512));
        } else {
            throw new RuntimeException("Not supported");
        }
    }
}

File: free-ethereum-core/src/main/java/org/ethereum/db/PeerSource.kt
/*
 * The MIT License (MIT)
 *
 * Copyright 2017 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 * Copyright (c) [2016] [ <ether.camp> ]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package org.ethereum.db

import org.apache.commons.lang3.tuple.Pair
import org.ethereum.datasource.DataSourceArray
import org.ethereum.datasource.ObjectDataSource
import org.ethereum.datasource.Serializer
import org.ethereum.datasource.Source
import org.ethereum.datasource.leveldb.LevelDbDataSource
import org.ethereum.net.rlpx.Node
import org.ethereum.util.RLP
import org.ethereum.util.RLPList
import org.slf4j.LoggerFactory
import java.math.BigInteger

/**
 * Source for [org.ethereum.net.rlpx.Node] also known as Peers
 */
class PeerSource(private val src: Source<ByteArray, ByteArray>) {
    var nodes: DataSourceArray<Pair<Node, Int>>? = null
        private set


    init {
        val INST = this
        this.nodes = DataSourceArray(
                ObjectDataSource(src, NODE_SERIALIZER, 512))
    }

    fun clear() {
        if (src is LevelDbDataSource) {
            src.reset()
            this.nodes = DataSourceArray(
                    ObjectDataSource(src, NODE_SERIALIZER, 512))
        } else {
            throw RuntimeException("Not supported")
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger("db")
        private val NODE_SERIALIZER = object : Serializer<Pair<Node, Int>, ByteArray> {

            override fun serialize(value: Pair<Node, Int>): ByteArray {
                val nodeRlp = value.left.rlp
                val nodeIsDiscovery = RLP.encodeByte(if (value.left.isDiscoveryNode) 1.toByte() else 0)
                val savedReputation = RLP.encodeBigInteger(BigInteger.valueOf(value.right.toLong()))

                return RLP.encodeList(nodeRlp, nodeIsDiscovery, savedReputation)
            }

            override fun deserialize(bytes: ByteArray?): Pair<Node, Int>? {
                if (bytes == null) return null

                val nodeElement = RLP.decode2(bytes)[0] as RLPList
                val nodeRlp = nodeElement[0].rlpData
                val nodeIsDiscovery = nodeElement[1].rlpData
                val savedReputation = nodeElement[2].rlpData
                val node = Node(nodeRlp)
                node.isDiscoveryNode = nodeIsDiscovery != null

                return Pair.of(node, if (savedReputation == null) 0 else BigInteger(1, savedReputation).toInt())
            }
        }
    }
}
